# Credit Risk Classification

In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df =  pd.read_csv('../Desktop/csv/lending_data.csv')

# Review the DataFrame
df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

In [5]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop(columns='loan_status')

In [10]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [9]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [15]:
# Check the balance of our target values
df.value_counts('loan_status')

loan_status
0    75036
1     2500
dtype: int64

In [16]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Create a Logistic Regression Model with the Original Data

In [19]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lr_df = LogisticRegression(random_state = 1)

# Fit the model using training data
lr_df.fit(X_train, y_train)

LogisticRegression(random_state=1)

In [20]:
# Make a prediction using the testing data
y_pred = lr_df.predict(X_test)

In [31]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_pred)

0.9918489475856377

In [32]:
confusion_matrix(y_test, y_pred)

array([[18663,   102],
       [   56,   563]])

In [34]:
# Print the classification report for the model
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Through the data given via the classification report it is evident to suggest that both 'healthy loan' and 'high risk-loan' are well predicted by the logisitic regression model, with accuracy values of 100% for healthy loans and 88% for high risk loans.

## Predict a Logistic Regression Model with Resampled Training Data

In [44]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_oversampled, y_oversampled = random_oversampler.fit_resample(X_train, y_train)

In [45]:
# Count the distinct values of the resampled labels data
y_oversampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

In [46]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lr_new_df = LogisticRegression(random_state = 1)

# Fit the model using the resampled training data
lr_new_df.fit(X_oversampled, y_oversampled)

# Make a prediction using the testing data
lr_pred = lr_new_df.predict(X_test)

In [49]:
# Print the balanced_accuracy score of the model 
accuracy_score(y_test, lr_pred)

0.9938093272802311

In [50]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, lr_pred)

array([[18649,   116],
       [    4,   615]])

In [51]:
# Print the classification report for the model
print(classification_report(y_test, lr_pred))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Given the data presented via the classification report, it is reasonable to assume that both 'healthy loan' and 'high risk-loan' are well predicted by the logisitic regression model, with accuracy values of 100% for healthy loans and 91% for high risk loans.